flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sink/AbstractWriterOperator.java

```diff
@@ -0,0 +1,147 @@
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.BoundedOneInput;
+import org.apache.flink.streaming.api.operators.InternalTimerService;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperator;
+import org.apache.flink.streaming.api.watermark.Watermark;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+
+import java.util.List;
+
+/**
+ * This is the abstract class used for executing the sink's {@link Writer} which is proposed by FLIP-143.
+ * The sub-class of this class is responsible for creating the specific {@link Writer} by implementing the {@link #createWriter()}.
+ *
+ * @param <InputT>   The type of writer's input
+ * @param <CommT>    The type of information needed to commit data staged by the sink
+ */
+@Internal
+abstract class AbstractWriterOperator<InputT, CommT> extends AbstractStreamOperator<CommT>
+	implements OneInputStreamOperator<InputT, CommT>, BoundedOneInput {
+
+	private static final long serialVersionUID = 1L;
+
+	/** The runtime information of the input element. */
+	private final Context<InputT> context;
+
+	// ------------------------------- runtime fields ---------------------------------------
+
+	/** We listen to this ourselves because we don't have an {@link InternalTimerService}. */
+	private Long currentWatermark;
+
+	/** The sink writer that does most of the work. */
+	private Writer<InputT, CommT, ?> writer;
+
+	AbstractWriterOperator() {
+		this.context = new Context<>();
+	}
+
+	@Override
+	public void open() throws Exception {
+		super.open();
+
+		this.currentWatermark = Long.MIN_VALUE;
+
+		writer = createWriter();
+	}
+
+	@Override
+	public void processElement(StreamRecord<InputT> element) throws Exception {
+		context.element = element;
+		writer.write(element.getValue(), context);
+	}
+
+	@Override
+	public void prepareSnapshotPreBarrier(long checkpointId) throws Exception {
+		super.prepareSnapshotPreBarrier(checkpointId);
+		sendCommittables(writer.prepareCommit(false));
+	}
+
+	@Override
+	public void processWatermark(Watermark mark) throws Exception {
+		super.processWatermark(mark);
+		this.currentWatermark = mark.getTimestamp();
+	}
+
+	@Override
+	public void endInput() throws Exception {
+		sendCommittables(writer.prepareCommit(true));
+	}
+
+	@Override
+	public void close() throws Exception {
+		super.close();
+		writer.close();
+	}
+
+	protected Sink.InitContext createInitContext() {
+		return new Sink.InitContext() {
+			@Override
+			public int getSubtaskId() {
+				return getRuntimeContext().getIndexOfThisSubtask();
+			}
+
+			@Override
+			public MetricGroup metricGroup() {
+				return getMetricGroup();
+			}
+		};
+	}
+
+	/**
+	 * @return A sink {@link Writer}
+	 * @throws Exception If creating {@link Writer} fail
+	 */
+	abstract Writer<InputT, CommT, ?> createWriter() throws Exception;
+
+	protected Writer<InputT, CommT, ?> getWriter() {
```

Comment:
For me I would more like removing `getWriter` method and directly make `writer` to be `protected`. 

Currently the override of this method in `StatefulWriterOperator` is only used in `snapshotState`, I think it would be ok for us to do type casting there.